src/opencl-poh-verify/cl_poh_verify.cpp

```diff
@@ -2,16 +2,178 @@
+                    g_total_gpus = 0;
+                    return false;
+                }
+            }
+        }
+    }
+    return g_total_gpus > 0;
+}
+
+gpu_ctx_t* get_gpu_ctx() {
+    int32_t cur_gpu, cur_queue;
+
+    LOG("getting gpu_ctx\n");
+
+    cur_gpu = g_cur_gpu;
+    g_cur_gpu++;
+    g_cur_gpu %= g_total_gpus;
+    cur_queue = g_cur_queue[cur_gpu];
+    g_cur_queue[cur_gpu]++;
+    g_cur_queue[cur_gpu] %= MAX_QUEUE_SIZE;
+
+    gpu_ctx_t* cur_ctx = &g_gpu_ctx[cur_gpu][cur_queue];
+    LOG("locking contex mutex queue: %d gpu: %d\n", cur_queue, cur_gpu);
+    pthread_mutex_lock(&cur_ctx->mutex);
+
+    LOG("selecting gpu: %d queue: %d\n", cur_gpu, cur_queue);
+
+    return cur_ctx;
+}
+
+void setup_gpu_ctx(gpu_ctx_t *cur_ctx,
+                    uint8_t* hashes,
+                    const uint64_t* num_hashes_arr,
+                    size_t num_elems,
+                    size_t nr_bytes_hashes,
+                    size_t nr_bytes_num_hashes_arr
+                   ) {
+	int ret;
+
+    LOG("device allocate. num hashes: %lu in MB nr_bytes_hashes: %f nr_bytes_num_hashes_arr: %f\n",
+        num_elems, (double)nr_bytes_hashes/(1024*1024), (double)nr_bytes_num_hashes_arr/(1024*1024));
+
+    if (cur_ctx->in_out_hashes == NULL || cur_ctx->in_num_elems < num_elems) {
+        clReleaseMemObject(cur_ctx->in_out_hashes);
+        cur_ctx->in_out_hashes = clCreateBuffer(context, CL_MEM_READ_WRITE, nr_bytes_hashes, NULL, &ret);
+        CL_ERR( ret );
+    }
+
+    if (cur_ctx->in_num_hashes_arr == NULL || cur_ctx->in_num_elems < num_elems) {
+        clReleaseMemObject(cur_ctx->in_num_hashes_arr);
+        cur_ctx->in_num_hashes_arr = clCreateBuffer(context, CL_MEM_READ_ONLY, nr_bytes_num_hashes_arr, NULL, &ret);
+        CL_ERR( ret );
+        cur_ctx->in_num_elems = num_elems;
+    }
+
+
+    CL_ERR( clEnqueueWriteBuffer(cmd_queue, cur_ctx->in_out_hashes, CL_TRUE, 0, nr_bytes_hashes, hashes, 0, NULL, NULL));
+    CL_ERR( clEnqueueWriteBuffer(cmd_queue, cur_ctx->in_num_hashes_arr, CL_TRUE, 0, nr_bytes_num_hashes_arr, num_hashes_arr, 0, NULL, NULL));
+}
+
+void release_gpu_ctx(gpu_ctx_t* cur_ctx) {
+    pthread_mutex_unlock(&cur_ctx->mutex);
+}
 
 extern "C" {
+
+void poh_verify_many_set_verbose(bool val) {
+    g_verbose = val;
+}
+
 int poh_verify_many(uint8_t* hashes,
                     const uint64_t* num_hashes_arr,
                     size_t num_elems,
                     uint8_t use_non_default_stream)
 {
-    fprintf(stderr, "poh_verify_many not implemented.");
-    exit(1);
+    if (num_elems == 0) return 0;
+
+    pthread_mutex_lock(&clg_ctx_mutex);
+    bool success = false;
+
+    if (initialized == false) {
+        success = cl_check_init();
+        DIE(success == false, "OpenCL could not be init");
+        DIE(cl_crypt_init_locked() == false, "cl_crypt_init_locked failed");
+        initialized = true;
+    } else {
+        fprintf(stderr, "cl_poh_verify_many already initialized\n");
```

Comment:
We aren't going to want this printing on every call.

Reply:
I replaced fprintf with LOG, if you want to remove it, I will. :+1:

Reply:
LOG is great.